Created New TUnit Aspire templates

licon4812 opened the pull request:
I have added two new template types to the template pack. This feature Resolves #1875


**TUnit.Aspire** which is based on the [.NET Aspire NUnit Template](https://learn.microsoft.com/en-us/dotnet/aspire/testing/write-your-first-test?pivots=nunit)

**TUnit.AspireStarter** is based on the .NET Aspire Starter template and creates the following project types. The projects are prefixed with Example.Namespace so that the projects will be prefixed with the user's namespace on project creation. Eg MyNamespace.TestProject, or MyNamespace.Apphost.

- ApiService
- AppHost
- ServiceDefaults
- TestProject
- WebApp

licon4812 commented:
The pending code quality issues are mainly default Aspire template stuff. How would you like me to proceed?

thomhurst commented:
> The pending code quality issues are mainly default Aspire template stuff. How would you like me to proceed?

That's fine - Don't worry about it.

thomhurst commented:
Wondering if we can come up with better names?

`TUnit.Aspire` and `TUnit.AspireStarter` feel too similar to me and don't clearly tell you which is which or what the difference is.

I also think an extra full stop would be better e.g. `TUnit.Aspire.Starter`

thomhurst commented on TUnit.Templates.Tests/Snapshots/InstantiationTest.TUnit.Aspire._.verified/TUnit.Aspire/TUnit.Aspire.csproj:
```diff
@@ -0,0 +1,24 @@
+    <Using Include="Microsoft.Extensions.DependencyInjection" />
+    <Using Include="Aspire.Hosting.ApplicationModel" />
+    <Using Include="Aspire.Hosting.Testing" />
+    <Using Include="TUnit" />
```
TUnit package does this so I don't think you need it here

thomhurst commented on TUnit.Templates.Tests/Snapshots/InstantiationTest.TUnit.AspireStarter._.verified/TUnit.AspireStarter/TUnit.AspireStarter.TestProject/GlobalSetup.cs:
```diff
@@ -0,0 +1,37 @@
+
+public class GlobalHooks
+{
+    public static DistributedApplication? App{ get; private set; }
```
If we know we're setting it in a before method we can avoid the nullability on the type. Also spacing issue with the brace.
```suggestion
    public static DistributedApplication App { get; private set; } = null!;
```

thomhurst commented on TUnit.Templates.Tests/Snapshots/InstantiationTest.TUnit.AspireStarter._.verified/TUnit.AspireStarter/TUnit.AspireStarter.TestProject/GlobalSetup.cs:
```diff
@@ -0,0 +1,37 @@
+public class GlobalHooks
+{
+    public static DistributedApplication? App{ get; private set; }
+    public static ResourceNotificationService? NotificationService { get; private set; }
```
```suggestion
    public static ResourceNotificationService NotificationService { get; private set; } = null!;
```

thomhurst commented on TUnit.Templates.Tests/Snapshots/InstantiationTest.TUnit.AspireStarter._.verified/TUnit.AspireStarter/TUnit.AspireStarter.TestProject/GlobalSetup.cs:
```diff
@@ -0,0 +1,37 @@
+        App = await appHost.BuildAsync();
+        NotificationService = App.Services.GetRequiredService<ResourceNotificationService>();
+        await App.StartAsync();
+    }
+
+    [After(TestSession)]
+    public static void CleanUp()
+    {
+        Console.WriteLine("...and after!");
+    }
```
```suggestion
    }
```

thomhurst commented on TUnit.Templates.Tests/Snapshots/InstantiationTest.TUnit.AspireStarter._.verified/TUnit.AspireStarter/TUnit.AspireStarter.TestProject/Hooks.cs:
```diff
@@ -0,0 +1,17 @@
+            Environment.SetEnvironmentVariable("PWDEBUG", "1");
+        }
+
+        Microsoft.Playwright.Program.Main(["install"]);
```
Copy and paste from the Playwright template? We don't need playwright here do we?

thomhurst commented on TUnit.Templates.Tests/Snapshots/InstantiationTest.TUnit.AspireStarter._.verified/TUnit.AspireStarter/TUnit.AspireStarter.TestProject/TUnit.AspireStarter.TestProject.csproj:
```diff
@@ -0,0 +1,30 @@
+
+
+  <ItemGroup>
+    <PackageReference Include="Aspire.Hosting.Testing" Version="9.0.0" />
+    <PackageReference Include="TUnit.Playwright" Version="0.13.0" />
```
```suggestion
    <PackageReference Include="Aspire.Hosting.Testing" Version="9.0.0" />
```

thomhurst commented on TUnit.Templates.Tests/Snapshots/InstantiationTest.TUnit.AspireStarter._.verified/TUnit.AspireStarter/TUnit.AspireStarter.TestProject/Tests/ApiTests.cs:
```diff
@@ -0,0 +1,41 @@
+    public async Task GetWeatherForecastReturnsOkStatusCode()
+    {
+        // Act
+        var httpClient = (GlobalHooks.App ?? throw new NullReferenceException()).CreateHttpClient("apiservice");
```
It's probably nicer to inject in the app via `[ClassDataSource<T>]` and make it implement `IAsyncInitializer` + `IAsyncDisposable` for the setup and teardown logic.

licon4812 replied:
This is the last thing I have left to do. Once I do this I will fix up the TUnit.Aspire.Starter template test to match

licon4812 commented:
> Wondering if we can come up with better names?
> 
> `TUnit.Aspire` and `TUnit.AspireStarter` feel too similar to me and don't clearly tell you which is which or what the difference is.
> 
> I also think an extra full stop would be better e.g. `TUnit.Aspire.Starter`

I have renamed `TUnit.Aspire` to `TUnit.Aspire.Test`. I have also fixed `TUnit.Aspire.Starter`.  As for the name of the starter template. It loosely matches the Aspire Starter template that the Aspire team provides.  Happy to update the naming in the template.config if you think it can match better.

![image](https://github.com/user-attachments/assets/a92f9ada-aeb5-4720-9cd9-216d03b4b807)
![image](https://github.com/user-attachments/assets/224bff9a-f62a-468f-af98-dfce5d58d88d)

licon4812 commented:
@thomhurst, I have fixed all the reviews you left. Feel free to let me know if there is anything else I need to fix. 

Thank you for the assistance with this 😊

licon4812 commented on TUnit.Templates/content/TUnit.Aspire.Test/IntegrationTest1.cs:
```diff
@@ -0,0 +1,36 @@
+        //     await using var app = await appHost.BuildAsync();
+        //     var resourceNotificationService = app.Services.GetRequiredService<ResourceNotificationService>();
+        //     await app.StartAsync();
+
```
Might be worth moving everything from the Arrange section to a ClassDataSource. Similar to how we did the TUnit.Aspire.Starter template

licon4812 commented:
Should be ready to go now @thomhurst

licon4812 commented:
@thomhurst auto merge turned off again :(
